contracts/adapters/UniswapAdapter.sol

```diff
@@ -0,0 +1,233 @@
+        return _name;
+    }
+
+    function nonFungible() external pure  returns (bool) {
+        return _nonFungible;
+    }
+
+    function sqrt(uint y) internal pure returns (uint z) {
+        if (y > 3) {
+            z = y;
+            uint x = y / 2 + 1;
+            while (x < z) {
+                z = x;
+                x = (y / x + x) / 2;
+            }
+        } else if (y != 0) {
+            z = 1;
+        }
+    }
+
+    // The core math involved in getting optimal swap amt to provide amm liquidity
+    function getSwapAmt(uint256 amtA, uint256 resA) internal pure returns (uint256){
+        return sqrt(amtA.mul(resA.mul(3988000) + amtA.mul(3988009))).sub(amtA.mul(1997))/1994;
+    }
+
+    function validateExchange(string memory exchangeName) internal view returns (bool){
+        for (uint i = 0; i < 2; i++){
+                if (keccak256(abi.encodePacked(exchangeName)) == keccak256(abi.encodePacked(venues[i]))){
+                        return true;
+                }
+        }
+        return false;
+    }
+
+    function convertEthToToken( uint256 inputAmount, address addr, string memory exchangeName) internal {
+        if (keccak256(abi.encodePacked(exchangeName)) == keccak256(abi.encodePacked(_uniswapExchange))){
+                _convertEthToToken(inputAmount, addr, uniswapRouter);
+        }
+        else if (keccak256(abi.encodePacked(exchangeName)) == keccak256(abi.encodePacked(_sushiswapExchange))){
+                _convertEthToToken(inputAmount, addr, sushiswapRouter);
+        }
+    }
+
+   function convertTokenToToken(address addr1, address addr2, uint256 amount, string memory exchangeName) internal {
+        if (keccak256(abi.encodePacked(exchangeName)) == keccak256(abi.encodePacked(_uniswapExchange))){
+                _convertTokenToToken(addr1, addr2, amount, uniswapRouter);
+        }
+        else if (keccak256(abi.encodePacked(exchangeName)) == keccak256(abi.encodePacked(_sushiswapExchange))){
+                _convertTokenToToken(addr1, addr2, amount, sushiswapRouter);
+        }
+   }
+
+   function addLiquidity(address token1,address token2,uint256 amount1,uint256 amount2,string memory exchangeName) internal {
+        if (keccak256(abi.encodePacked(exchangeName)) == keccak256(abi.encodePacked(_uniswapExchange))){
+                _addLiquidity(token1, token2, amount1, amount2, uniswapRouter);
+         }
+
+         else if (keccak256(abi.encodePacked(exchangeName)) == keccak256(abi.encodePacked(_sushiswapExchange))){
+                _addLiquidity(token1, token2, amount1, amount2, sushiswapRouter);
+         }
+
+      }
+
+   function _convertEthToToken( uint256 inputAmount, address addr, IUniswapV2Router02 router) internal {
+        uint deadline = block.timestamp + 150;
+        address[] memory path = new address[](2);
+        path[0] = wethAddress;
+        path[1] = addr;
+        uint256 expectedOutput = router.getAmountsOut(inputAmount, path)[1];

```

Comment:
We cannot safely rely on `getAmountsOut` for the price due to a [sandwich attack](https://uniswap.org/docs/v2/smart-contract-integration/trading-from-a-smart-contract/#safety-considerations).

The fix here is to let the user pass in the `amountOutMin` into the function